File: src/main/java/org/noxet/noxetserver/menus/inventory/BlockedPlayersMenu.java
package org.noxet.noxetserver.menus.inventory;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.ClickType;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.UsernameStorageManager;
import org.noxet.noxetserver.menus.ItemGenerator;
import org.noxet.noxetserver.menus.chat.ChatPromptMenu;
import org.noxet.noxetserver.playerdata.PlayerDataManager;
import org.noxet.noxetserver.util.InventoryCoordinate;
import org.noxet.noxetserver.util.InventoryCoordinateUtil;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;

public class BlockedPlayersMenu extends InventoryMenu {
    private final List<String> blockedPlayers;
    private final InventoryCoordinate blockNewPlayerSlot;

    public BlockedPlayersMenu(Player player) {
        super((new PlayerDataManager(player).getListSize(PlayerDataManager.Attribute.BLOCKED_PLAYERS) - 1) / 9 + 2, "Blocked Players", false);
        //noinspection unchecked
        blockedPlayers = (List<String>) new PlayerDataManager(player.getUniqueId()).get(PlayerDataManager.Attribute.BLOCKED_PLAYERS);

        blockNewPlayerSlot = InventoryCoordinateUtil.getCoordinateFromXY(0, getInventory().getSize() / 9 - 1);
    }

    @Override
    protected void updateInventory() {
        for(String blockedUUIDString : blockedPlayers) {
            UUID blockedUUID = new UsernameStorageManager().getUUIDFromUsernameOrUUID(blockedUUIDString);
            String blockedName = UsernameStorageManager.getCasedUsernameFromUUID(blockedUUID);

            setSlotItem(
                    ItemGenerator.generatePlayerSkull(
                            NoxetServer.getPlugin().getServer().getOfflinePlayer(blockedUUID),
                            "§c" + (blockedName != null ? blockedName : blockedUUIDString),
                            Arrays.asList(
                                    "§7UUID: " + blockedUUIDString,
                                    "§e→ Double-click to §a§npardon§e."
                            )
                    ),
                    InventoryCoordinateUtil.getCoordinateFromSlotIndex(blockedPlayers.indexOf(blockedUUIDString))
            );
        }

        setSlotItem(
                ItemGenerator.generateItem(
                        Material.PAPER,
                        1,
                        "§aBlock Player",
                        Arrays.asList("§7Block a player to", "§7prevent them from sending", "§7messages and friend/TPA requests."),
                        true
                ), blockNewPlayerSlot
        );
    }

    @Override
    protected boolean onSlotClick(Player player, InventoryCoordinate coordinate, ClickType clickType) {
        if(coordinate.isAt(blockNewPlayerSlot)) {
            new ChatPromptMenu("player to block", player, promptResponse -> {
                player.performCommand("block " + promptResponse.getMessage());
                new BlockedPlayersMenu(player).openInventory(player);
            });
            return true;
        }

        if(clickType != ClickType.DOUBLE_CLICK)
            return false;

        if(coordinate.getSlotIndex() > blockedPlayers.size() - 1)
            return false;

        String clickedBlockedUUID = blockedPlayers.get(coordinate.getSlotIndex());

        player.performCommand("unblock " + clickedBlockedUUID);

        new BlockedPlayersMenu(player).openInventory(player);

        return true;
    }
}

File: src/main/java/org/noxet/noxetserver/menus/inventory/BlockedPlayersMenu.scala
package org.noxet.noxetserver.menus.inventory

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.menus.ItemGenerator
import org.noxet.noxetserver.menus.chat.ChatPromptMenu
import org.noxet.noxetserver.playerdata.PlayerDataManager
import org.noxet.noxetserver.util.{InventoryCoordinate, InventoryCoordinateUtil, UsernameStorageManager}

import java.util.{Arrays, UUID}

class BlockedPlayersMenu(player: Player) extends InventoryMenu(
  (PlayerDataManager(player).getListSize(PlayerDataManager.Attribute.BlockedPlayers) - 1) / 9 + 2,
  "❌ Blocked Players",
  false):

  private val blockedPlayers = PlayerDataManager(player.getUniqueId).get(PlayerDataManager.Attribute.BlockedPlayers).asInstance[List[String]]
  private val blockNewPlayerSlot = InventoryCoordinateUtil.getCoordinateFromXY(0, getInventory.getSize / 9 - 1)

  override protected def updateInventory(): Unit =
    for blockedUuidString <- blockedPlayers do
      val blockedUuid = UsernameStorageManager().getUuidFromUsernameOrUuid(blockedUuidString)
      val blockedName = UsernameStorageManager.getCasedUsernameFromUuid(blockedUuid)

      setSlotItem(
        ItemGenerator.generatePlayerSkull(
          NoxetServer.getPlugin.getServer.getOfflinePlayer(blockedUuid),
          s"§c${blockedName.getOrElse(blockedUuidString)}",
          Some(List(
            s"§7UUID: $blockedUuidString",
            "§e→ Double-click to §a§npardon§e."
          ))
        ),
        InventoryCoordinateUtil.getCoordinateFromSlotIndex(blockedPlayers.indexOf(blockedUuidString))
      )

    setSlotItem(
      ItemGenerator.generateItem(
        Material.PAPER,
        "§aBlock Player",
        Some(List(
          "§7Block a player to",
          "§7prevent them from sending",
          "§7messages and friend/TPA requests."
        ))
      ), blockNewPlayerSlot
    )

  override protected def onSlotClick(player: Player, coordinate: InventoryCoordinate, clickType: ClickType): Boolean =
    if coordinate.isAt(blockNewPlayerSlot) then
      ChatPromptMenu("player to block", player, promptResponse => {
        player.performCommand(s"block ${promptResponse.getMessage}")
        BlockedPlayersMenu(player).openInventory(player)
      })
      return true

    if clickType != ClickType.DOUBLE_CLICK || coordinate.slotIndex > blockedPlayers.size - 1 then
      return false

    val clickedBlockedUuid = blockedPlayers.get(coordinate.slotIndex)

    player.performCommand(s"unblock $clickedBlockedUuid")

    BlockedPlayersMenu(player).openInventory(player)

    true